src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt

```diff
@@ -0,0 +1,203 @@
+        return CargoCheckAnnotationInfo(file, editor)
+    }
+
+    override fun doAnnotate(info: CargoCheckAnnotationInfo) = getCachedResult(info.file).value
+
+    override fun apply(file: PsiFile, annotationResult: CargoCheckAnnotationResult?, holder: AnnotationHolder) {
+        annotationResult ?: return
+        val doc = holder.currentAnnotationSession.file.viewProvider.document ?: throw AssertionError()
+
+        for (topMessage in annotationResult.messages) {
+            val message = topMessage.message
+            val filePath = holder.currentAnnotationSession.file.virtualFile.path
+
+            val severity = when (message.level) {
+                "error" -> HighlightSeverity.ERROR
+                "warning" -> HighlightSeverity.WEAK_WARNING
+                else -> throw AssertionError()
+            }
+
+            val codeStr = message.code?.code
+
+            // If spans are empty we add a "global" error
+            if (message.spans.isEmpty()) {
+                if (topMessage.target.src_path == filePath) {
+                    // add a global annotation
+                    val annotation = holder.createAnnotation(severity, TextRange.EMPTY_RANGE, message.message)
+                    annotation.isFileLevelAnnotation = true
+                    annotation.setNeedsUpdateOnTyping(true)
+                    annotation.tooltip = escapeHtml(message.message) + if (codeStr != "") "<hr/>$codeStr" else ""
+                }
+            } else {
+                val problemGroup = ProblemGroup { message.message }
+
+                val relevantSpans = message.spans.filter {
+                    val spanFilePath = PathUtil.getCanonicalPath(it.file_name)
+                    filePath.endsWith(spanFilePath) && isValidSpan(it)
+                }
+
+                for (span in relevantSpans) {
+                    val annotation = createAnnotation(span, message, severity, doc, holder)
+                    annotation.problemGroup = problemGroup
+                    annotation.setNeedsUpdateOnTyping(true)
+                }
+            }
+        }
+    }
+
+    fun createAnnotation(span: CargoSpan, message: CargoMessage, severity: HighlightSeverity, doc: Document,
+                         holder: AnnotationHolder): Annotation {
+
+        fun toOffset(line: Int, column: Int): Int {
+            val lineStart = doc.getLineStartOffset(line)
+            return lineStart + column
+        }
+
+        // The compiler message lines and columns are 1 based while intellij idea are 0 based
+        val textRange =
+            TextRange(toOffset(span.line_start - 1, span.column_start - 1),
+                toOffset(span.line_end - 1, span.column_end - 1))
+
+        val shortMessage =
+            when (span.is_primary) {
+                true -> message.message
+                else -> span.label ?: message.message
+            }
```

Comment:
It would be nice to add an error code (if it's available) to this message, as we do in error annotators: "can't find crate for `foo` [E0463]"